### DISCUSSION TITLE
Line draw start at shifted x position + proposed solution

### DISCUSSION THREAD

**Discussion by itzzp** [highfestiva/finplot#330] [General]
Hello Highfestiva,

Thank you for the fantastic plotting tool!

When drawing a line with ctrl&mousedrag I would expect the line to start at the position of the crosshair, but it is shifted a bit along the x-axis. Is this intentional? 
By changing the text on line 810 and 817 in the__init__.py file from mapToView to mapSceneToView this is solved. That feels more logical to me.

Kind regards,
Herman

**Comment by itzzp** [comment]
Excuse me: that should be line 788 and 792

**Reply by highfestiva** [reply]
Ah, that may well be the case. Could you post a small example where the line drawing doesn't work?

**Comment by itzzp** [comment]
Hello,

You can reproduce it by using your own **dockable.py** example (https://github.com/highfestiva/finplot/blob/master/finplot/examples/dockable.py). 

**After line 37 add: 
ax0.showGrid(x=True, y=True, alpha=0.1)**

Then run and **draw a line** by holding the ctrl-button and the mousebutton and dragging the mouse. **The line will start shifted to the right.**  

Kind regards,
Herman

**Reply by highfestiva** [reply]
Interesting. My new line segement click happens way over to the left. It didn't help for me to replace with `mapSceneToView()`. I'll investigate further.

**Reply by itzzp** [reply]
Thank you for your time.

**Reply by highfestiva** [reply]
Spent some time on it, and cobbled together a fix in 9c8674b. The root cause is unknown though, but probably in pyqtgraph.

**Reply by itzzp** [reply]
Thank you!

**Comment by itzzp** [comment]
Hello Highfestiva,

I checked-out the latest version and saw the changes you've made to circumvent the issue. Thank you for that, ,but I think the bug is still  there.

You can still reproduce it by using your own dockable.py example (https://github.com/highfestiva/finplot/blob/master/finplot/examples/dockable.py).

After line 37 add:
ax0.showGrid(x=True, y=True, alpha=0.1)

Then run and draw a line by holding the ctrl-button and the mousebutton and dragging the mouse. 

In the previous 'bug report' I said the line starts to the right but after the changes the line starts **1 window size** too far to the left (at least that's what I think after playing around with it). The example starts with the ticker AMRK, the price is around 80 dollar at the moment. If you, just for the experiment, start drawing the line just a bit to the right of the y-axis label '80', you see the line appear on the left side of the screen. So for the experiment, start drawing on the extreme right side of dock_0.

I hope this helps you find the bug.

Have a good weekend.
Kind regards,
Herman

**Reply by highfestiva** [reply]
I don't get that bug. Could you send me a gif showing the behavior and how to reproduce? How did you "install" the finplot update?

**Comment by itzzp** [comment]
Hello Highfestiva,

Yes I installed the update.

But: When drawing a line by CTRL + Left-mouse-drag the line is shifted a window size to the left. In your function mapClickToView(self, pos) (line 883) you shift the position width() to the right to prevent this (line 888). The function mapClickToView is used in mouseClickEvent (line 868) so when I add a point to a line by clicking this is correct. But, and I think this is still missing, this is not used in mouseLeftDrag (line 792) when I start a new line.

In short:
These changes solve the problem for me.

Line 806 is:  p1 = self.mapToView(ev.pos())
Line 806 should be: p1 = self.mapClickToView(ev.pos())  

Line 810 is:  p0 = self.mapToView(ev.lastPos())
Line 810 should be: p0 = self.mapClickToView(ev.lastPos()) 

Hope this helps, if not please say so, I will try to create a gif but I do not yet know how.

Kind regards,
Herman

**Comment by itzzp** [comment]
![ctrl_left_drag_01](https://user-images.githubusercontent.com/773814/170825578-772a81c0-7e3e-4635-8b2e-c2212eefb1db.gif)

**Comment by itzzp** [comment]
Sorry for the spam but it is maybe best to tackle these related issues at once.

The same bug is there with right drag.

Example dockable.py
I added one line: 
ax0.showGrid(x=True, y=True, alpha=0.1)


Right drag in dock_1 does show a ‘zoom area’ at the correct position


Right drag in dock_0 does not show a ‘zoom area’ at the correct position

You can see the window appear on the left when the mouse is on the right

![right_drag](https://user-images.githubusercontent.com/773814/170948811-6474dc68-ff2a-4fa2-b217-709afa3e350e.gif)

**Reply by highfestiva** [reply]
Super strange, works for me. How did you install finplot in the first place, and how did you install the update?

**Comment by itzzp** [comment]
Yes indeed that's strange. I installend the __init__.py file by downloading the zip-file from github page: https://github.com/highfestiva/finplot (Code -> download zip). Unpack the zip and paste the  init file in a directory of a pycharm project where I use it. But if you do not have the same error i might try to reinstall pyqtgraph or pyQt. Are you sure you added the line to draw gridlines in the example code? If I leave it out it works flawless.

**Reply by highfestiva** [reply]
What version of pyqt5 and pyqtgraph are you using?

**Reply by highfestiva** [reply]
These are mine:
```
PyQt5                        5.15.6
PyQt5-Qt5                    5.15.2
PyQt5-sip                    12.9.1
pyqtgraph                    0.12.4
PyQtWebEngine                5.15.2
```

**Comment by itzzp** [comment]
PyQt5                            5.15.6
PyQt5-Qt5                     5.15.2
PyQt5-sip                     12.9.0
pyqtgraph                     0.12.3

I will update PyQt5-sip and pyqtgraph

**Comment by itzzp** [comment]
Well I could have tried that before. With:
PyQt5-sip-12.10.1 and 
pyqtgraph-0.12.4
it is working. Sorry ..

**Reply by highfestiva** [reply]
Interesting! Don't know what the sip module does. If you ever take the time to downgrade one of them, you can determine which module contains the bug. Sadly this will probably return again (when they fix the bug on their side, our quickfix will create a bug on our side). When that happens, just let me know and I'll quickfix it back. :-/
